pkg/monitor/main.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"strconv"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/cilium/cilium/api/v1/models"
+	"github.com/cilium/cilium/common"
+	"github.com/cilium/cilium/daemon/defaults"
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/byteorder"
+	"github.com/cilium/cilium/pkg/monitor/payload"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+const targetName = "cilium-node-monitor"
+
+func receiveEvent(m *bpf.PerfEventSample, c int) {
+	pl := payload.Payload{Data: m.DataCopy(), CPU: c, Lost: 0, Type: payload.EventSample}
+	send(pl)
+}
+
+func lostEvent(m *bpf.PerfEventLost, c int) {
+	pl := payload.Payload{Data: []byte{}, CPU: c, Lost: m.Lost, Type: payload.RecordLost}
+	send(pl)
+}
+
+var (
+	listeners     = make(map[int]net.Conn)
```

Comment:
A map seems inefficient here. I would use a different data structure. You seem to need a data structure that makes it cheap to add elements, allows iterating, and easily removing elements while iterating. I would recommend using a list:

    listeners = list.New()

All your operations would be O(1).